Feature: admin notifications

SpeedyD opened the pull request:
Got @preimpression's permission to port this, and so far that is purely that. Still gonna look coding through to see if things are necessary.. like, I think "UserCharacterLog" is not necessary for this? Gonna have to test that..

Also, need to still add toggles for users and overall, since this is somewhat obtrusive.

Also thinking of just making the circles badges? I think the huge circle is a bit much and bootstrap's badges are perfect for this.. Opinions on that welcome.

preimpression committed:
Adds an admin notification bubble (similar to on deviantart) that lists the number of items in the queue. Staff with "manage_masterlist" will get to see trades, transfers, and design queues. Staff with "manage_submissions" will see claim and prompt queue.

preimpression committed:
Merge branch 'master' into adminNotification

preimpression committed:
Merge branch 'corowne:master' into adminNotification

preimpression committed:
feat: Update to LK 2.1.x and adjust name

preimpression committed:
refactor: Adjusts way User model collects admin notifications and adds Reports to the notifications.

SpeedyD committed:
Merge remote-tracking branch 'preimpression/extension/admin-notification' into feature/admin-notifications

* preimpression/extension/admin-notification:
  refactor: Adjusts way User model collects admin notifications and adds Reports to the notifications.
  Adds an admin notification bubble (similar to on deviantart) that lists the number of items in the queue. Staff with "manage_masterlist" will get to see trades, transfers, and design queues. Staff with "manage_submissions" will see claim and prompt queue.

SpeedyD committed:
refactor: fix blade formatting

SpeedyD committed:
refactor: fix PHP styling

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

itinerare commented:
This should have toggles before being merged, imo.

SpeedyD committed:
Major updates, will explain in comment

SpeedyD committed:
refactor: fix PHP styling

SpeedyD committed:
Forgot credits. Should've really done this sooner.

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

perappu commented on app/Models/User/User.php:
```diff
@@ -18,8 +20,10 @@
+use App\Models\Report\Report;
 use App\Models\Shop\ShopLog;
 use App\Models\Submission\Submission;
+use App\Models\Trade;
```
```suggestion
use App\Models\Trade\Trade;
```

Needs to be updated to this due to https://github.com/lk-arpg/lorekeeper/pull/1199

SpeedyD replied:
Right, wasn't aware. Thanks.

perappu commented on app/Models/User/User.php:
```diff
@@ -722,4 +726,37 @@ public function getSubmissions($user = null) {
+     *
+     * @return int
+     */
+    public function hasAdminNotification($user) {
```
I hate to be like this, but this function creates an _extremely_ gnarly amount of queries if you're on a user that has a good amount of permissions (and this is only a sampling, AND it's on the admin account so it's not making all of the rank-related DB calls that an ordinary user would have). It's a problem I've always had with the original extension. And since this appears in the navbar, it's on _every_ page:

<img width="3660" height="1492" alt="Image" src="https://github.com/user-attachments/assets/669da003-40f6-416f-b9a9-2d3a55153fb6" />

Is there.... any way to make this better? Like at all? I feel like this is starting to get into some of the real flaws we have with hasPower() at the moment...

SpeedyD replied:
Boy howdy, that's a good question. Since this is straight up a port from the ext, I haven't actually thought about many changes.. But you are right.. perhaps something needs to be done.

..what if it saves the numbers somewhere, like in a new database table, pull that instead.. and.. Hrm.. it would be updated if.. anyone submits something? Might mean a lot of new functions spread around..

...I'd have to think about it. Suggestions can always be PR'd to the https://github.com/SpeedyD/lorekeeper/tree/feature/admin-notifications branch itself, haha.. 😅

perappu replied:
If you get a chance, can you update this branch to latest dev to grab the hasPower() changes from https://github.com/lk-arpg/lorekeeper/pull/1479? It won't solve everything, but I'd be curious to pull it again and see if the queries were at all improved from that alone.

SpeedyD replied:
Updated to latest dev. Only change I had to enforce was removing
`use App\Models\Rank\RankPower;`
`use App\Models\Report\Report;`

I don't exactly have the time or mental space rn to fix up the queries. I once again would like to ask/suggest perhaps submitting a PR to the above branch. ^^;

perappu replied:
No no that's exactly why I asked, I'd like to look at it myself & see if it improved just from the latest core updates alone. Thanks for grabbing that!

SpeedyD replied:
In that case, no problem at all, haha. Thanks for looking into it. :)

perappu commented on app/Models/User/User.php:
```diff
@@ -722,4 +726,37 @@ public function getSubmissions($user = null) {
+        $count[] = $this->hasPower('manage_reports') ? Report::where('status', 'Pending')->count() : 0; // reportCount
+        $count[] = $this->hasPower('manage_reports') ? Report::assignedToMe($this)->count() : 0; // assignedReportCount
+
+        // If Adoption Center is installed:
```
I'll be honest, I'm not 100% sure how I feel about adding accommodations for specific extensions in core, even if it was in the original extension. I think this is the only time in the entire codebase that we'd be doing so, so it feels a bit odd... I don't really foresee either of the ones here making it into core either.

itinerare replied:
Not per se; the command for migrating subtypes to the new format that supports multiple has handling for the couple of multiple-subtypes commands. That's not so much "standing support" as "support to migrate away from", though....

perappu replied:
That's true, and when I thought about it more we do have that 'criteria' in ParsePostRequestFields, but that extension is imminently coming into core (and has been a candidate for core for a very long time), so...

itinerare replied:
Yeah... that one's been a notable exception on the grounds of intended inclusion in core for a while.

SpeedyD replied:
Mrh, something to consider, I suppose, but until we can sort out the new query handling, I think I'll just let this one sit.

SpeedyD replied:
..thinking about it now, I feel like once this is actually _in_ core, it would the _extension developer's_ responsibility to add this, not ours. So I am going to remove this.

SpeedyD committed:
Update app/Models/User/User.php

Co-authored-by: Kazimir <stokoriel@gmail.com>

SpeedyD committed:
Merge branch 'develop' into feature/admin-notifications

SpeedyD committed:
Removed extension accomodations

For those with adoption center and affiliates, please request the extension developer to add to this function. (Once it's added to core.)

liwoyadan committed:
feat(admin notifs): cache the notif, also reduce rank power queries

- also the use line for reports seemed to be missing so i added that in

liwoyadan committed:
fix(admin notifs): wrong variable name for reports whoops

SpeedyD committed:
Merge pull request #217 from liwoyadan/feature/admin-notifications

feat(admin notifs): cache the notif, also reduce rank power queries
